### PR TITLE
Prefer memcpy and memcmp over direct memory read/comparisons.

### LINKED ISSUE #126: SSE alignment warnings
```
zlib-ng/arch/x86/insert_string_sse.c:26:15: runtime error: load of misaligned address 0x6310000a0801 for type 'unsigned int', which requires 4 byte alignment
0x6310000a0801: note: pointer points here
 00 80 11  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00
              ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/arch/x86/insert_string_sse.c:26:15 in 
zlib-ng/match.c:363:36: runtime error: load of misaligned address 0x6310000a0803 for type 'uint16_t' (aka 'unsigned short'), which requires 2 byte alignment
0x6310000a0803: note: pointer points here
 11  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00
              ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/match.c:363:36 in 
zlib-ng/match.c:412:13: runtime error: load of misaligned address 0x6310000a0801 for type 'uint16_t' (aka 'unsigned short'), which requires 2 byte alignment
0x6310000a0801: note: pointer points here
 00 80 11  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00
              ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/match.c:412:13 in 
zlib-ng/match.c:427:32: runtime error: load of misaligned address 0x6310000a0804 for type 'unsigned long', which requires 8 byte alignment
0x6310000a0804: note: pointer points here
  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
              ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/match.c:427:32 in 
zlib-ng/match.c:428:32: runtime error: load of misaligned address 0x6310000a0803 for type 'unsigned long', which requires 8 byte alignment
0x6310000a0803: note: pointer points here
 11  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00
              ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/match.c:428:32 in 
zlib-ng/match.c:398:17: runtime error: load of misaligned address 0x6310000ab2ed for type 'uint16_t' (aka 'unsigned short'), which requires 2 byte alignment
0x6310000ab2ed: note: pointer points here
 00 00 00 04 0c 75 ff  04 0d 76 ff 04 0c 75 ff  03 0b 75 ff 03 0b 75 ff  03 0b 75 ff 03 0b 75 ff  03
             ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/match.c:398:17 in 
zlib-ng/match.c:453:24: runtime error: load of misaligned address 0x6310000ab2f7 for type 'uint16_t' (aka 'unsigned short'), which requires 2 byte alignment
0x6310000ab2f7: note: pointer points here
 ff 04 0c 75 ff  03 0b 75 ff 03 0b 75 ff  03 0b 75 ff 03 0b 75 ff  03 0b 74 ff 03 0b 74 ff  03 0b 74
             ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/match.c:453:24 in 
zlib-ng/match.c:362:36: runtime error: load of misaligned address 0x6310000ab30d for type 'uint16_t' (aka 'unsigned short'), which requires 2 byte alignment
0x6310000ab30d: note: pointer points here
 0b 74 ff 03 0b 74 ff  03 0b 74 ff 03 0b 74 ff  03 0b 74 ff 03 0b 73 ff  03 0b 73 ff 03 0b 73 ff  03
             ^ 
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/match.c:362:36 in 
zlib-ng/inflate.c:1167:17: runtime error: left shift of 198 by 24 places cannot be represented in type 'int'
SUMMARY: AddressSanitizer: undefined-behavior zlib-ng/inflate.c:1167:17 in 
```

I got this when running a large batch of unit tests which use `zlib-ng`. Only some of them produce alignment issues.

### PR DESCRIPTION
**Pull request by ioquatix**
* Some platforms have alignment requirements and unaligned direct memory
  read/comparisons may result in undefined behaviour.
* Prefer memcpy and memcmp which are lowered to efficient assembly where
  possible.

Please review carefully to ensure I've not made any mistakes with order of arguments or other oversights.

Fixes #126

### CODE REVIEW

**Comment by ioquatix** [comment]
I see some issues like this:

```
/home/travis/build/Dead2/zlib-ng/memcopy.h:24:5: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
     uint32_t res = chunk.a;
     ^
/home/travis/build/Dead2/zlib-ng/memcopy.h: In function ‘set_bytes’:
/home/travis/build/Dead2/zlib-ng/memcopy.h:193:9: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
         unsigned char c = *from;
         ^
/home/travis/build/Dead2/zlib-ng/memcopy.h: In function ‘chunk_memcpy’:
/home/travis/build/Dead2/zlib-ng/memcopy.h:230:5: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
     unsigned rem = len % sz;
     ^
/home/travis/build/Dead2/zlib-ng/memcopy.h:235:5: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
     unsigned by8 = len % sz;
     ^
/home/travis/build/Dead2/zlib-ng/memcopy.h: In function ‘byte_memset’:
/home/travis/build/Dead2/zlib-ng/memcopy.h:290:5: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
     unsigned char *from = out - 1;
     ^
/home/travis/build/Dead2/zlib-ng/memcopy.h:295:5: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
     unsigned rem = len % sz;
     ^
/home/travis/build/Dead2/zlib-ng/memcopy.h:300:5: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
     unsigned by8 = len % 8;
     ^
/home/travis/build/Dead2/zlib-ng/memcopy.h: In function ‘chunk_memset’:
/home/travis/build/Dead2/zlib-ng/memcopy.h:359:5: warning: ISO C90 forbids mixed declarations and code [-Wpedantic]
     unsigned sz = sizeof(uint64_t);
```

What is your preference? I can move the variables before `memcpy`.

File: match.c [hunk 1]
```diff
@@ -361,9 +361,11 @@ ZLIB_INTERNAL unsigned longest_match(deflate_state *const s, IPos cur_match) {
     unsigned int wmask = s->w_mask;
 
     register unsigned char *strend = s->window + s->strstart + MAX_MATCH;
-    register uint16_t scan_start = *(uint16_t*)scan;
-    register uint16_t scan_end   = *(uint16_t*)(scan+best_len-1);
-
+
```
Comment (mtl1979) [inline match.c]: This empty line should be after "uint16_t scan_start, scan_end;" due to C requirement that all declarations should be before any of the variables are used. The empty line after variable declarations makes it clear where declarations stop and code starts.

File: match.c [hunk 2]
```diff
@@ -426,8 +428,9 @@ ZLIB_INTERNAL unsigned longest_match(deflate_state *const s, IPos cur_match) {
         scan += 2, match+=2;
         Assert(*scan == *match, "match[2]?");
         do {
-            unsigned long sv = *(unsigned long*)(void*)scan;
-            unsigned long mv = *(unsigned long*)(void*)match;
+            unsigned long sv, mv;
+            memcpy(&sv, scan, sizeof(sv)); // = *(unsigned long*)(void*)scan;
+            memcpy(&mv, match, sizeof(mv)); // = *(unsigned long*)(void*)match;
```
Comment (mtl1979) [inline match.c]: This code is invalid in C... memcpy should be after "unsigned long xor" is declared.

**Review by mtl1979 — CHANGES_REQUESTED** [review]

File: match.c (hunk 1 above)
Comment (ioquatix) [reply match.c]: Makes sense, I just wanted to keep the relevant memcpy next to the variable, but I guess in strict C we can't organise code like that.

However, there are other issues in the code where variables and code is mixed. I will fix this one.

File: match.c [hunk 3]
```diff
@@ -428,10 +428,11 @@ ZLIB_INTERNAL unsigned longest_match(deflate_state *const s, IPos cur_match) {
         scan += 2, match+=2;
         Assert(*scan == *match, "match[2]?");
         do {
-            unsigned long sv, mv;
-            memcpy(&sv, scan, sizeof(sv)); // = *(unsigned long*)(void*)scan;
-            memcpy(&mv, match, sizeof(mv)); // = *(unsigned long*)(void*)match;
-            unsigned long xor = sv ^ mv;
+            unsigned long sv, mv. xor;
+            memcpy(&sv, scan, sizeof(sv));
```
Comment (mtl1979) [inline match.c]: One empty line after all declarations.

Comment (ioquatix) [reply match.c]: Excellent, fixed.

File: match.c [hunk 4]
```diff
@@ -426,9 +429,12 @@ ZLIB_INTERNAL unsigned longest_match(deflate_state *const s, IPos cur_match) {
         scan += 2, match+=2;
         Assert(*scan == *match, "match[2]?");
         do {
-            unsigned long sv = *(unsigned long*)(void*)scan;
-            unsigned long mv = *(unsigned long*)(void*)match;
-            unsigned long xor = sv ^ mv;
+            unsigned long sv, mv. xor;
```
Comment (mtl1979) [inline match.c]: Should be "unsigned long sv, mv, xor;"
There is period instead of comma between "mv" and "xor".

Comment (ioquatix) [reply match.c]: How did this slip through? how did it even compile :(

**Comment by ioquatix** [comment]
Okay, issues should all be sorted, sorry, the period was too much, I rebased :)

**Comment by ioquatix** [comment]
I also removed trailing whitespace on the lines I added.

**Comment by ioquatix** [comment]
Okay thanks for being patient with me and explaining feedback clearly. I appreciate that :) Let me know if there is anything else that requires attention.

**Comment by mtl1979** [comment]
@ioquatix At quick glance I can't see anything else right now...

**Comment by Dead2** [comment]
Regarding the C89/C90 warnings, I was a little surprised to see that we have not defined C99 anywhere. Because I distinctly remember doing that at some point, possibly last year or even before that. So I guess that commit must have gotten lost in a branch somewhere along the way.

I believe we should get this done now though, because if we require C99, it will not really make much problems for compilers except the need for some workarounds for MSVC (and we already do that). And it is required for function inlining that we use for example. Also support for the C++ comments is something I'd like to make more use of, for obvious reasons.

So fixing those warnings is not a priority, since it is valid C99 code.

Also, I can't get travis to do checks now, seems their macos test servers are unable to do dns lookups of "github.com".

**Comment by ioquatix** [comment]
Can you point me towards the relevant part of the C99 spec which makes this defined behaviour?

**Comment by Dead2** [comment]
@ioquatix I am talking about the C89 warnings you posted "ISO C90 forbids mixed declarations and code" and the removal of c++ style comments. I am not talking about the memcpy/memcmp stuff.

The memcpy/memcmp stuff is something we'll have to look into a bit more thoroughly, but it does look good so far.

**Comment by ioquatix** [comment]
Okay, sorry I did misunderstand you but you've clarified it perfectly. Thanks.

**Comment by ioquatix** [comment]
Any update to this?

**Comment by mtl1979** [comment]
There hasn't been much discussion lately... We have been busy with other tasks... We will post update when there is something new...

**Comment by ioquatix** [comment]
Okay great. Is there anything else I can do?

**Comment by mtl1979** [comment]
@ioquatix I'm pretty sure there is a lot people can do, but it makes no sense to do much until @Dead2 has time to go through the backlog of pull requests.

**Comment by ioquatix** [comment]
Fair enough, okay I'll wait.

**Comment by Dead2** [comment]
Regarding my progress, the renamelib branch has been used in production on 50+ servers (with nginx only, so a quite limited scope) the last couple of months now, with zero reported issues and a nice little speed boost.
My current hope is to re-work the renamelib branch next week, and prepare it for merging into the main branch. After that, it'll be a lot easier to accept and test PRs.
Sorry for being slow with updates, but I have been increasingly sick this past year. I am hoping 2018 is the turning point. I do read every comment left on github (via email notifications).

**Comment by ioquatix** [comment]
@Dead2 I'm sorry to hear about your health. If there is anything I can do to help let me know.

**Comment by ioquatix** [comment]
Thanks so much for your effort @Dead2 it's great that this has been merged. If there is anything else I can help with let me know. :) Hope you are feeling well.

**Pull request merged**